packages/loopring_v3/contracts/amm/libamm/AmmJoinProcess.sol

```diff
@@ -83,109 +37,178 @@ library AmmJoinProcess
         )
         internal
     {
-        S.validatePoolTransaction(
+        S.checkPoolTxApproval(
             join.owner,
-            AmmJoinRequest.hashPoolJoin(ctx.domainSeparator, join),
+            AmmJoinRequest.hash(ctx.domainSeparator, join),
             signature
         );
 
+        bool joinFromLayer1 = (
+            join.direction == AmmData.Direction.L1_TO_L1 ||
+            join.direction == AmmData.Direction.L1_TO_L2
+        );
+
+        if (joinFromLayer1) {
+            require(signature.length == 0, "LAYER1_JOIN_WITH_OFFCHAIN_APPROVAL_DISALLOWED");
+
+            if (join.nonce == S.joinLocks[msg.sender].length) {
+                S.joinLocks[msg.sender].pop();
+            } else {
+                delete S.joinLocks[msg.sender][join.nonce - 1];
+
+                if (S.joinLockIdx[msg.sender] == join.nonce - 1) {
+                    S.joinLockIdx[msg.sender]++;
+                }
+            }
```

Comment:
`msg.sender` -> `join.owner`.

Seems like a risk currently is that `joinLocks` can grow without limit for users, and for withdrawing we have to run over this list, so could potentially lock up user funds if the list is too big.

Could you explain `joinLockIdx`? Or maybe the use will explain itself when withdrawals are done.